yarnpkg/yarn
# Install transitive dependencies even if listed as devDependencies in the project

Fixes #2819

Go to #2819 for a minimal way to reproduce it. Basically, if you have these kind of dependency tree:
`package.json`:
```
{
  dependencies: {
    "a": "1.0.0"
  },
  devDependencies: {
    "b": "1.0.0"
  }
}
```

`a/package.json`:
```
{
  dependencies: {
    "b": "1.0.0"
  }
}
```

If you install your package using the `--production` flag, then `b` won't be installed (and it should, since it's a transitive dependency of `a`). This PR fixes this.

The issue was that, if the exact same version of a dependency pattern is found multiple times during the `resolve` phase, only the first time it will actually be resolved. In order to prune `devDependencies`, all packages that have exactly 1 "resolve request" after the `resolve` phase will be discarded. I've removed the optimisation that checks if a given pattern has already been resolved, and changed an array to a `Set` to not impact performance.

This has probably not been a good explanation but I'm not very familiar with this codebase. I've tried to see what was the motivation for that optimisation, but it's there since the creation of this repo.

**Test plan**

* Create a project like the one described in #2819, with `sequelize-cli` as a dependency and `gulp` as a devDependency.
* Run `yarn install --production`
* In `master`, `gulp` won't be present in `node_modules`. With this PR, it will be.
* In `master`, `yarn check verify-tree` will fail. With this PR it will pass.
